util/chunk/recordbatch.go

```diff
@@ -13,12 +13,41 @@
 
 package chunk
 
+// UnspecifiedNumRows represents requiredRows is not specified.
+const UnspecifiedNumRows = -1
+
 // RecordBatch is input parameter of Executor.Next` method.
 type RecordBatch struct {
 	*Chunk
+
+	// requiredRows indicates how many rows is considered full for parent executor.
+	// Child executor can return immediately if there are such number of rows,
+	// instead of fulling the whole chunk.
+	// This is not compulsory, so the number of returned rows can be larger than it in some cases.
+	requiredRows int
 }
 
 // NewRecordBatch is used to construct a RecordBatch.
 func NewRecordBatch(chk *Chunk) *RecordBatch {
-	return &RecordBatch{chk}
+	return &RecordBatch{chk, UnspecifiedNumRows}
+}
+
+// SetRequiredRows sets the number of rows the parent executor want.
+func (rb *RecordBatch) SetRequiredRows(numRows int) *RecordBatch {
+	if numRows <= 0 {
+		numRows = UnspecifiedNumRows
+	}
+	rb.requiredRows = numRows
+	return rb
+}
+
+// RequiredRows returns how many rows the parent executor want.
+func (rb *RecordBatch) RequiredRows() int {
+	return rb.requiredRows
+}
+
+// IsFull returns if this batch can be considered full.
+func (rb *RecordBatch) IsFull(maxChunkSize int) bool {
```

Comment:
can we remove `maxChunkSize` from the interface? Only check whether `rb.NumRows() >= rb.requiredRows`? we should set `rb.requiredRows` according to max chunk size?

Reply:
Only checking `requiredRows` is dangerous and not convenient.
I prefer to take both `requiredRows` and `MaxChunkSize` into account in this function.
How about introducing `sessionctx.Context` into `RecordBatch`.
Then the `IsFull` will behave like:
```
IsFull() bool {
    numRows >= ctx.GetSessionVars().MaxChunkSize || 
    (requiredRows != UnspecifiedNumRows && numRows >= requiredRows)
}
```

Reply:
Is `requiredRows` determined by `maxChunkSize`?

Reply:
`requiredRows` should be equal or less than `maxChunkSize`, but it can be `Unspecified`.
When it's `Unspecified`, it's dangerous to always return `true` if it ignores `maxChunkSize`.

And in each place where `IsFull` is used, it is not convenient to check both `IsFull` and `maxChunkSize` like:
```
if !batch.IsFull() && batch.NumRows() < ctx.GetSessionVars().MaxChunkSize { ... }
```

Reply:
I found there are some executors have their own `maxChunkSize` instead of using global `maxChunkSize` in `sessionctx.Context`, so the approach mentioned above is not appropriate.

I will remove `maxChunkSize` from `IsFull` and let the outside scope where `IsFull` is called check it.